Wikipedia/Code/WMFSearchViewController.m

```diff
@@ -523,22 +523,29 @@ - (void)saveLastSearch {
 
 #pragma mark - Languages
 
-- (NSArray*)allLanguagesFromController {
-    NSMutableArray* lang = [NSMutableArray array];
-    [lang addObjectsFromArray:[MWKLanguageLinkController sharedInstance].preferredLanguages];
-    [lang addObjectsFromArray:[MWKLanguageLinkController sharedInstance].otherLanguages];
-    return lang;
-}
-
 - (void)updateLanguages {
-    NSArray* languages = [self allLanguagesFromController];
+    NSArray* languages = [MWKLanguageLinkController sharedInstance].preferredLanguages;
     self.searchLanguages = [languages wmf_arrayByTrimmingToLength:3];
 }
 
 - (void)updateLanguageButtonsToPreferredLanguages {
     [self updateLanguages];
-    [self.languageButtons enumerateObjectsUsingBlock:^(UIButton* _Nonnull obj, NSUInteger idx, BOOL* _Nonnull stop) {
-        [obj setTitle:[(MWKLanguageLink*)self.searchLanguages[idx] localizedName]  forState:UIControlStateNormal];
+    [self.searchLanguages enumerateObjectsUsingBlock:^(MWKLanguageLink*  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
```

Comment:
seems like you're enumerating twice, with each enumeration referencing the other array. could these enumerations be consolidated into one for simplicity? :

``` objc
// all we care about is configuring the buttons, so enumerate those:
[self.languageButtons enumerateObjecstsUsingBlock:^(UIButton* obj, NSUInteger idx, BOOL* _) {
  if (idx < self.searchLanguages.count) { 
    obj.hidden = NO;
    [obj setTitle:self.searchLanguages[idx].localizedName];    
  } else {
    // hide buttons which are out of bounds
    obj.hidden = YES 
  }
}];
```

I also omitted setting `enabled` in my example since hidden buttons should be disabled automatically.